src/sqlfluff/utils/reflow/depthmap.py

```diff
@@ -0,0 +1,128 @@
+"""The DepthMap class which is an enriched sequence of raw segments."""
+
+import logging
+from dataclasses import dataclass
+from typing import List, Sequence, Set, Tuple
+
+from sqlfluff.core.parser import BaseSegment
+from sqlfluff.core.parser.segments.raw import RawSegment
+
+
+reflow_logger = logging.getLogger("sqlfluff.rules.reflow")
+
+
+@dataclass(frozen=True)
+class DepthInfo:
+    """An object to hold the depth information for a specific raw segment."""
+
+    stack_depth: int
+    stack_hashes: List[int]
+    # This is a convenience cache to speed up operations.
+    stack_hash_set: Set[int]
+    stack_class_types: List[Set[str]]
+
+    @classmethod
+    def from_raw_and_stack(cls, raw: RawSegment, stack: Sequence[BaseSegment]):
+        """Construct from a raw and it's stack."""
+        stack_hashes = [hash(seg) for seg in stack]
+        return cls(
+            stack_depth=len(stack),
+            stack_hashes=stack_hashes,
+            stack_hash_set=set(stack_hashes),
+            stack_class_types=[seg.class_types for seg in stack],
+        )
+
+    def common_with(self, other: "DepthInfo") -> List[int]:
+        """Get the common depth and hashes with the other."""
+        # We use set intersection because it's faster and hashes should be unique.
+        common_hashes = self.stack_hash_set.intersection(other.stack_hashes)
+        common_depth = len(common_hashes)
+        return self.stack_hashes[:common_depth]
```

Comment:
Is there an assumption here that `self` and `other` share a common prefix in terms of `stack_hashes`?

A doctest example, relevant assertion, or a bit more description could be helpful here. (Or can you think of a more descriptive function name?)

Reply:
Yes there is, although I should probably make that more explicit. In theory they should at least share a `FileSegment` parent.

Reply:
Cool. My motivation for this comment is 90% about making the intent easier for humans to understand, and 10% about catching possible future errors.